### ISSUE TITLE
[proxy] Increase default outbound TCP loadshed threshold / queue capacity (or provide annotation to do so)

### ISSUE THREAD

**Issue by euanhunteratom** [linkerd/linkerd2#11565]
### What problem are you trying to solve?

Increase the default outbound TCP loadshed threshold, which I believe is the outbound TCP queue capacity from [here](https://github.com/linkerd/linkerd2-proxy/blob/main/linkerd/app/src/env.rs#L487). The current default is 10 which seems very small, something like 100 or 1000 seems more reasonable (HTTP queue capacity is default 10k for comparison).

The usecase we have is that our application lazily allocates DB connections. If there is a burst of requests it is very easy for our application to try and open more than 10 TCP connections to the remote database at once. When this happens any attempt over the 10th fails with an IO error, the Linkerd proxy container logs the following messages, and the `outbound_tcp_errors_total` metric increases with the reason "loadshed"
```
INFO ThreadId(01) outbound: linkerd_app_core::serve: Connection closed error=service unavailable client.addr=X.X.X.X:X
```

We are currently working around this issue by using `skip-outbound-port` to have the proxy skip the database port since it is external to the service mesh and so doesn't need to be proxied. However we are concerned about this behaviour if we ever move the database in-mesh, or how this might impact other TCP or HTTPS (since Linkerd treats that as TCP) traffic within the mesh. So, either changing this default or allowing it to be changed via an annotation would be great!

### How should the problem be solved?

Change [DEFAULT_OUTBOUND_TCP_QUEUE_CAPACITY](https://github.com/linkerd/linkerd2-proxy/blob/cbf226e8e5e8fe7d8c3378b3b346f381ef8d7832/linkerd/app/src/env.rs#L247) to something higher like 100 or 1000.

The drawback to this would be the potential for higher memory consumption by the proxy. Given, however, that the queue capacity for HTTP requests is 10k, 10 vs 100 is likely to be a negligible change in memory consumption.

### Any alternatives you've considered?

An alternative would be a new annotation recognised by the controller that would set `ENV_OUTBOUND_TCP_QUEUE_CAPACITY` on the injected proxy sidecar. This would give the user a lot of control over the queue capacity, allowing the default to remain and only be increased by the user when they feel the need. It does, however, add complexity to the controller.

### How would users interact with this feature?

Since this is a default setting, they wouldn't.

If the annotation approach was taken, the user would interact with it via a Pod annotation as in [Proxy Configuration](https://linkerd.io/2.14/reference/proxy-configuration/).

### Would you like to work on this feature?

maybe

**Comment by DavidMcLaughlin** [comment]
Thanks for reporting this - we agree this is probably too low. We want to do some scale testing to figure out exactly how the queue capacity impacts resource usage and will then update this ticket. Either way we'll at least make this configurable via an annotation.

**Comment by olix0r** [comment]
@euanhunteratom https://github.com/linkerd/linkerd2-proxy/commit/bf5cb67b0acfc0f100ddaf41ceb5d3cbde7913b3 was merged to address this issue. This setting change is available in the edge-23.11.3 proxy. You can test this proxy version by setting the pod annotation `config.linkerd.io/proxy-version: edge-23.11.3`. I expect that we'll include this change in the next stable-2.14 backport.

I agree, though, that we should ideally expose this setting as configuration. I didn't want that design to get in the way of fixing the obviously bad default behavior.

**Comment by dkulchinsky** [comment]
Hey @olix0r, I wonder if there's a target version of linkerd to include this fix?

We're on stable-2.14.8 which comes with linkerd-proxy 2.210.4, but from what I can tell the fix is available from linkerd-proxy 2.213.0 (https://github.com/linkerd/linkerd2-proxy/releases/tag/release%2Fv2.213.0)

I generally find it a bit tricky to figure out which linkerd-proxy version is included in each linkerd release 😅 maybe there's some place where this is tracked?

**Comment by alpeb** [comment]
> I generally find it a bit tricky to figure out which linkerd-proxy version is included in each linkerd release 😅 maybe there's some place where this is tracked?

Right, it's not obvious. You'd have to select the release tag in the linkerd2 repo and then see the file `.proxy-version`.
